### ISSUE TITLE
Komodo IDE 11 - SCM shortcuts and actions don't always work (might be related to path length)

### ISSUE THREAD

**Issue by daigo75** [Komodo/KomodoEdit#3314]
Copied from original post: https://community.komodoide.com/t/komodo-ide-11-scm-key-bindings-dont-always-work/3965

### Short Summary
In Komodo IDE 11, the key bindings set for the SCM (Git, in this case), but they don't seem to work as they should. I opened a file within a Git repository, made a change to it, then tried to commit the change using the shortcuts I configured, but nothing happens. The same applies to the Push function: I used the using the correct key bindings, the prompt "waiting for command" appears, then I press the last key and nothing happens.

I tried performing the same operations from the menu, with mixed results. Sometimes, the commit was unavailable on a single file, which I had just modified, but it could be performed on the folder.

It might be a red herring, but I have the feeling that the path might have something to do about the issue. Examples:

* If I open `C:\Users\MyUserName\Documents\Projects\Web\customers\Applications\Modules\Ecommerce\My_Project_Name\my-custom-project-with-long-name\my-custom-project-with-long-name.php`, where `my-custom-project-with-long-name` is a Git repository, the hotkeys don't work. The commit and push operations don't happen (no popup window).
* If I open the file as `C:\Temp\my-custom-project-with-long-name\my-custom-project-with-long-name.php`, then the hotkeys (both push and commit) work.
For completeness, I performed the same test with an older version of Komodo IDE, to ensure that the issue is not related to the Git tools on the machine, and the hotkeys work in both the above scenarios. This looks like a bug in this latest version (also found in KIDE 10).

### Steps to Reproduce
1. Open a file within a Git repository, with a long path (mine was 170 characters).
2. Change the file.
3. Save the file. The name becomes red (which, I presume, means that it has changed from the repo version).
4. Use the hotkeys to perform a commit. Nothing happens.
5. Try to perform the commit from the context menu. It may or may not work.
6. If the commit works, use the hotkeys to perform a push. Nothing happens.

If you repeat the above with the same file, but with a shorter path name, the hotkeys and related operations work.

### Expected results
The hotkeys should trigger the commit/push operations.

### Actual results
The hotkeys don't work. :smile: 

### Platform Information
* Editor: Komodoe IDE 11
* OS: Windows 7 Professional

### Additional Information
Relevant extract from the error log, when the hotkeys are used:
```
  [2017-10-21 17:58:40,023] [ERROR] console-logger: NS_ERROR_FAILURE: Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [koISCCDVCS.getValue] (2) in chrome://komodo/content/scc/scc.js:481
    Traceback (most recent call last):
      File "chrome://komodo/content/scc/scc.js", line 481, in 

  [2017-10-21 17:58:40,026] [ERROR] console-logger: NS_ERROR_FAILURE: Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [koISCCDVCS.getValue] (2) in chrome://komodo/content/scc/scc.js:481
    Traceback (most recent call last):
      File "chrome://komodo/content/scc/scc.js", line 481, in 
   
  [2017-10-21 17:58:40,190] [ERROR] scc.history: 
    -- EXCEPTION START --
    [Exception... "Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [koISCCDVCS.getValue]"  nsresult: "0x80004005 (NS_ERROR_FAILURE)"  location: "JS frame :: chrome://komodo/content/scc/scc.js :: this.getRepositoryRoot :: line 481"  data: no]
    + toString (function) 3 lines
    + message (string) 'Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [koISCCDVCS.getValue]'
    + result (number) 2147500037
    + name (string) 'NS_ERROR_FAILURE'
    + filename (string) 'chrome://komodo/content/scc/scc.js'
    + lineNumber (number) 481
    + columnNumber (number) 0
    + inner (object) null
    | + data (object) null
    | + stack (string) 'this.getRepositoryRoot@chrome://komodo/content/scc/scc.js:481:19
    scc_history_OnLoad@chrome://komodo/content/scc/history.js:185:20
    observer.observe@chrome://komodo/content/scc/history.js:840:16
    '
    + location (object) JS frame :: chrome://komodo/content/scc/scc.js :: this.getRepositoryRoot :: line 481
    | + QueryInterface (function) 3 lines
    | + language (number) 2
    | + languageName (string) 'JavaScript'
    | + filename (string) 'chrome://komodo/content/scc/scc.js'
    | + name (string) 'this.getRepositoryRoot'
    | + lineNumber (number) 481
    | + columnNumber (number) 19
    | + sourceLine (string) ''
    | + caller (object) JS frame :: chrome://komodo/content/scc/history.js :: scc_history_OnLoad :: line 185
    | + formattedStack (string) 'this.getRepositoryRoot@chrome://komodo/content/scc/scc.js:481:19
    scc_history_OnLoad@chrome://komodo/content/scc/history.js:185:20
    observer.observe@chrome://komodo/content/scc/history.js:840:16
    '
    | + stack
        this.getRepositoryRoot@chrome://komodo/content/scc/scc.js:481:19
        scc_history_OnLoad@chrome://komodo/content/scc/history.js:185:20
        observer.observe@chrome://komodo/content/scc/history.js:840:16
    -- EXCEPTION END --
```

**Comment by Naatan** [comment]
Thanks for the report!

**Comment by cgchoffman** [comment]
This appears to be a **Git+Windows** issue, not a Komodo issue.  In attempting to setup an env to fix this I'm unable to get any further than creating a dir with a long name.  When I try to clone a repo into that dir Git fails:
```
C:\Users\cgcho\Projects\TEMP\aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa>git clone git@github.com:cgchoffman/testing-3314.git
Cloning into 'testing-3314'...
fatal: insanely long object directory C:/Users/cgcho/Projects/TEMP/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/testing-3314/.git/objects
```

I found various references to this issue on the internet: https://github.com/git-for-windows/git/issues/342#issuecomment-220891129

My guess is that @daigo75 was able to clone a long repo name onto his machine but Git fails on various commands ones that "long object name" is on Windows and Git is trying to work with it.

**Comment by daigo75** [comment]
That might be, but then how comes that the exact same issue doesn't occur with an old version of Komodo IDE? I "exhumed" the installer of a trial version of Komodo IDE 7, and it works fine. The shortcuts open the Git dialog, the commit and push work as expected, etc., even with the long file names that cause issues in Komodo 11.

Perhaps older versions of Komodo relied on a different mechanism to interface with Git?

**Comment by cgchoffman** [comment]
I forgot about that part of your report @daigo75, thanks for pointing it out again.  I'll have another look.

**Comment by cgchoffman** [comment]
I still can't reproduce this.  I'm able to get a very long path+repo name onto my machine but Komodo never fails.  Maybe Git updated things so you can't get into this situation?  I can't reproduce this bug at all.

- I created a new repo with as long a name as github allows (physically stops you from making it longer): `testing-3314-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa`
- I made as long a path as I could before cloning that repo.  Any longer and I would get the `insanely long object directory` error when cloning
- Once the repo is cloned locally I can edit the readme in Komodo
- *ctrl+K,C*, dialog opens, I commit
- *ctrl+k,p*, push dialog opens,  I push

Everything works fine in Komodo 11.0.2.

Any chance you can double check you're able to reproduce this issue @daigo75?

**Comment by cgchoffman** [comment]
FYI, i have this repo for quick testing: https://github.com/cgchoffman/testing-3314-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa

**Comment by daigo75** [comment]
I'm afraid that I can't use Komodo 11 anymore. The trial period expired a while ago, I don't have a licence for it yet.

**Comment by cgchoffman** [comment]
@daigo75 I understand.  

Note to self: Try making path to repo longer after cloning

**Comment by daigo75** [comment]
Please note that the repository name was not that long, in my case. It was full file name that was long. You can create a folder like `C:\Users\MyUserName\Documents\Projects\Web\customers\Applications\Modules\Ecommerce\My_Project_Name\my-custom-project-with-long-name`, run `Git init` inside it and add a file with a long name. That might cause the issue.

**Comment by cgchoffman** [comment]
@daigo75, thanks i'll try that too.

**Comment by cgchoffman** [comment]
Tested Komodo 7, 8, 11 and git in cmd with the same repo.  I created a file with a very long filename and added it to the repo.  I then extended path the repo is in until Windows wouldn't let me anymore.  This just breaks Git across the board, whether it's in the terminal or in Komodo.  Once I climb back up the folders enough, the repo works in all contexts.

I just can't reproduce the scenario you describe @daigo75.  I'm going to close this ticket.  You're effort here has been GREATLY appreciated though.  Thank you for chiming back in so quickly to try to help me out.
